[#7598] Port Perspective Broker to Python 3

glyph opened the pull request.

hawkowl committed:
try this instead

hawkowl committed:
some syntax porting

hawkowl committed:
Merge branch 'oldstyle-bases-8308' into port-pb-7598

hawkowl committed:
checkpoint

hawkowl committed:
checkpoint

hawkowl committed:
checkpoint

hawkowl committed:
checkpoint

hawkowl committed:
cleanup

hawkowl committed:
checkpoint

hawkowl committed:
make the pb tests pass

hawkowl committed:
checkpoint

hawkowl committed:
move the tests into a better spot

hawkowl committed:
checkpoint

hawkowl committed:
checkpoint

hawkowl committed:
checkpoint

hawkowl committed:
checkpoint

hawkowl committed:
Merge remote-tracking branch 'origin/trunk' into port-pb-7598

hawkowl committed:
merge forward

hawkowl committed:
changes

hawkowl committed:
fix... something

glyph committed:
more b's

glyph committed:
even more b's

glyph committed:
i love b's

glyph committed:
duplicate import

glyph committed:
_accurately_ copy InstanceType behavior

glyph committed:
refactor all instance-creation code

put the stuff that is used by Cacheable into flavors, on Cacheable,
rather than making it look general purpose; funnel all 2/3 instance
creation stuff through one function

glyph committed:
more modules

glyph committed:
test modules

glyph committed:
log.err()

glyph committed:
various pyflakes people

glyph committed:
topfile

glyph committed:
Merge branch 'trunk' into port-pb-7598

glyph committed:
reverse out old-style decorator changes

glyph committed:
remove "_range" refactoring

the code here was already future-proof, if not ideally efficient; kill
these lines in the interest of making the diff shorter.

glyph committed:
add test_pb and test_jelly to py3 manifest

glyph committed:
eliminate unnecessary dependence on _EXPECT_NEWSTYLE

glyph committed:
more diff reduction

glyph committed:
fix

glyph committed:
include all the modules please

glyph committed:
and why not

hawkowl commented on twisted/python/dist3.py:
```diff
@@ -299,6 +299,8 @@
     "twisted.scripts.twistd",
     "twisted.spread.__init__",
     "twisted.spread.banana",
```
I think we need 'flavors' here too...

glyph replied:
yep, added

glyph committed:
this test is kind of nonsense even on py2

it's totally meaningless on py3

glyph committed:
we already have skip computations for these

glyph committed:
no longer need this

glyph committed:
preserving 'long'-ness is not actually the protocol's job

glyph committed:
removing more random cleanups

glyph committed:
Merge branch 'trunk' into port-pb-7598

glyph committed:
fewer lines

glyph committed:
can't cheat and make the lines too long

fewer lines is winning but at what cost

glyph committed:
fewer lines

glyph committed:
remove unnecessary cleanup

glyph committed:
deal with comparability the easy way

glyph committed:
reverse another unnecessary cleanup

glyph committed:
keep the relative imports

this was on a changed line, but let's not regress back towards top-level
imports; if we do clean stuff up, we should clean up towards explicit
relative intra-package imports everywhere

glyph committed:
remove more cleanups

glyph committed:
new lines shouldn't be >80

glyph committed:
unrelated

glyph committed:
*ForLocal can handle unicode so let them do it

generally idiomatic usage will be native strings with these methods
anyway, so let's not change the tests.

glyph committed:
uncleanup

glyph committed:
aesthetics; fewer blanks

glyph committed:
as long as we _do_ have to modify every line here

let's at least make it shorter

glyph committed:
also unnecessary

glyph committed:
also unrelated

glyph committed:
oops

glyph committed:
put back the old way of treating the _sets module

glyph committed:
reduce diff size some more

glyph committed:
reverse another irrelevant change

glyph committed:
Merge branch 'trunk' into port-pb-7598

glyph committed:
Merge remote-tracking branch 'origin/port-pb-7598' into port-pb-7598

hawkowl commented:
Py3.3 doesn't seem to be happy...

glyph commented:
The only test I see failing on 3.3 is `twisted.trial.test.test_tests.UnhandledDeferredTests.test_doesntBleed` which is completely unrelated as far as I can tell.  It is a garbage collection thing; maybe Travis's infra is experiencing some memory pressure?

hawkowl commented:
Considering its failing on push _and_ pr builders (which are separate job
runs) I find that unlikely... Did the GC semantics change in 3.4? And do we
care if 3.3 doesn't work? (Apart from pypy 3 I guess)

On 29 Jul 2016 03:26, "Glyph" notifications@github.com wrote:

> The only test I see failing on 3.3 is
> twisted.trial.test.test_tests.UnhandledDeferredTests.test_doesntBleed
> which is completely unrelated as far as I can tell. It is a garbage
> collection thing; maybe Travis's infra is experiencing some memory pressure?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/twisted/twisted/pull/436#issuecomment-235999186, or mute
> the thread
> https://github.com/notifications/unsubscribe-auth/ADJ2XHyF6WgbBZZ7WOhEhjTqPdGiqZWoks5qaQJtgaJpZM4JW0uU
> .

glyph commented:
I have no idea. That test should be totally unrelated!

hawkowl commented:
Are the pb tests leaving a reference around, maybe? Does it happen on 3.4
if you run the tests in the same order as 3.3?

On 29 Jul 2016 05:19, "Glyph" notifications@github.com wrote:

> I have no idea. That test should be totally unrelated!
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/twisted/twisted/pull/436#issuecomment-236028610, or mute
> the thread
> https://github.com/notifications/unsubscribe-auth/ADJ2XBooRqVm7kTCu2Wqw_zXue4n5lsEks5qaRzogaJpZM4JW0uU
> .

glyph commented:
I suppose I should run 3.3 locally first.  But if I were going to - how would I run them in the same order?

adiroiban commented on twisted/cred/checkers.py:
```diff
@@ -76,7 +76,7 @@ class InMemoryUsernamePasswordDatabaseDontUse(object):
 
     def __init__(self, **users):
-        self.users = users
+        self.users = {x.encode('ascii'):y for x, y in users.items()}
```
minor comment:

why not encode with utf-8 ? ... so that we can also use this code to test unicode related usernames / passwords ?

adiroiban commented on twisted/spread/flavors.py:
```diff
@@ -16,33 +16,35 @@
-from .jelly import Jellyable, Unjellyable, _newDummyLike
-from .jelly import setInstanceState, getInstanceState
+from twisted.python.compat import _PY3, unicode, comparable, cmp
+from .jelly import (
```
are relative import a thing we want to use in Twisted?

just trying to understand the coding standard

adiroiban commented on twisted/spread/flavors.py:
```diff
@@ -107,6 +109,10 @@ def remoteMessageReceived(self, broker, message, args, kw):
         kw = broker.unserialize(kw)
+
+        if not isinstance(message, str):
+            message = message.decode('utf8')
```
this code is not covered ... and I find it hard to understand why we got it here.

I am expected that remote messages are always received as bytes... so they should always be decoded.

Can we have the `message` argument documented to make it clear what types are accepted ?

adiroiban commented on twisted/spread/jelly.py:
```diff
@@ -88,12 +90,15 @@
         # Python 3
         _sets = None
+    else:
+        _SetTypes.append(_sets.Set)
```
minor comment: can these code raise an ImportError, why not move it closer to the `import sets` ... I find it easier to read

adiroiban commented on twisted/spread/pb.py:
```diff
@@ -981,14 +1005,16 @@ def proto_error(self, requestID, fail):
         """
+        if isinstance(objectID, unicode):
+            objectID = objectID.encode('utf8')
         refs = self.localObjects[objectID].decref()
```
since localObjects is a public API, maybe it should be replaced with a dict like implementation which converts the objectID on the fly . I guess that external code might also want to access it and the forget to do the conversion..

or maybe it should be private and it needs to be deprecated as a public API

adiroiban replied:
I saw there are many other places in which we convert the key for localObjects. With that in mind I think that a dict like class is needed.

With manual conversion we might forget to convert at some place and then we end up with errors.

adiroiban commented on twisted/spread/pb.py:
```diff
@@ -882,6 +903,9 @@ def _recvMessage(self, findObjMethod, requestID, objectID, message, answerRequir
         invoke it with args, and send an 'answer' or 'error' response.
         """
+        if not isinstance(message, str):
+            message = message.decode('utf8')
```
why decoding here since remoteMessageReceived will also do the decoding ?

adiroiban commented on twisted/spread/pb.py:
```diff
@@ -557,12 +564,17 @@ def expressionReceived(self, sexp):
             command = sexp[0]
+
+            if not isinstance(command, str):
+                command = command.decode('utf8')
```
can expressionReceived be called with non-bytes commands ?

adiroiban commented on twisted/spread/jelly.py:
```diff
@@ -889,18 +915,9 @@ def _unjelly_instance(self, rest):
-            self.postCallbacks.append(inst.postUnjelly)
-        return inst
+        if not _PY3 and type(clz) is not _OldStyleClass:
+            raise InsecureJelly("Legacy 'instance' found with new-style class")
```
I don't see any tests for this change ... if we are testing just for InsecureJelly I think that we should also assert the InsecureJelly's message

adiroiban commented:
I have re-run the 3.3 job on travis-ci and it still fails ...  while the tests seems unrelated, maybe the new spread tests are having some side effects .

I have re-run it on trunk on travis-ci and it pass.

I guess there is something wrong about the current changes :)

glyph commented:
Thanks for the review comments @adiroiban .  I will look into the py3.3 stuff...

rodrigc commented:
I can confirm that twisted.trial.test.test_tests.UnhandledDeferredTests.test_doesntBleed is consistently failing under Python 3.3.
I put a print() in the test to print out the value of **gc.garbage** inside the test and found:

```
[<twisted.spread.test.test_pb.DumbPub object at 0x108544650>, <twisted.spread.pb.RemoteReference object at 0x109f15110>, <twisted.spread.test.test_pb.DumbPub object at 0x109ec0290>, <twisted.spread.pb.RemoteReference object at 0x108533f90>, <twisted.spread.pb.RemoteReference object at 0x108e465d0>, <twisted.spread.pb.RemoteReference object at 0x10848ec10>, <twisted.spread.pb.RemoteReference object at 0x10848ed10>, <twisted.spread.pb.RemoteReference object at 0x10848ed50>, <twisted.spread.pb.RemoteReference object at 0x10848ebd0>, <twisted.spread.pb.RemoteReference object at 0x10851c150>, <twisted.spread.pb.RemoteReference object at 0x10851c5d0>, <twisted.spread.pb.RemoteReference object at 0x10851e850>, <twisted.spread.pb.RemoteReference object at 0x1082e1950>, <twisted.spread.pb.RemoteReference object at 0x1082e1e90>, <twisted.spread.pb.RemoteReference object at 0x1082e1250>, <twisted.spread.pb.RemoteReference object at 0x10850bc10>, <twisted.spread.pb.RemoteReference object at 0x10850bf10>, <twisted.spread.pb.RemoteReference object at 0x10850be90>, <twisted.spread.pb.RemoteReference object at 0x108f73490>, <twisted.spread.pb.RemoteReference object at 0x10842a290>, <twisted.spread.pb.RemoteReference object at 0x10842a250>, <twisted.spread.pb.RemoteReference object at 0x108422110>, <twisted.spread.pb.RemoteReference object at 0x108422710>, <twisted.spread.pb.RemoteReference object at 0x10a57f850>, <twisted.spread.pb.RemoteReference object at 0x10a57fbd0>, <twisted.spread.pb.RemoteReference object at 0x10a57fed0>, <twisted.spread.pb.RemoteReference object at 0x10a57f7d0>, <twisted.spread.pb.RemoteReference object at 0x10a57fb10>, <twisted.spread.pb.RemoteReference object at 0x10a57fad0>, <twisted.spread.pb.RemoteReference object at 0x10a57f9d0>, <twisted.spread.pb.RemoteReference object at 0x10a57fdd0>, <twisted.spread.pb.RemoteReference object at 0x10a57f4d0>, <twisted.spread.pb.RemoteReference object at 0x10a57f450>, <twisted.spread.pb.RemoteReference object at 0x10a57f910>, <twisted.spread.pb.RemoteReference object at 0x10a57f510>,
```

with over 1000 entries of twisted.spread.pb.RemoteReference object.

It looks like this is triggering behavior which was changed in Python 3.4 under https://www.python.org/dev/peps/pep-0442/

glyph commented:
@rodrigc Thanks for investigating and confirming.

glyph commented:
The link to the PEP was particularly useful :)

glyph committed:
merge trunk

glyph committed:
address garbage collection issues

test failures on 3.3 are the result of some weird flaw in the garbage
collector which pollutes gc.garbage and causes later tests to fail; not
present on later versions of Python.

This change, however, breaks any errant circular references by
triggering connectionLost, and eliminates unnecessary Deferreds and
associated round-trip traffic in the streaming producer portions of PB
so that tests don't spuriously fail.

glyph committed:
don't disconnect if we're already disconnected

rodrigc commented:
👍 This looks good.  I tested it on my system with Python 3.3.6, and could not reproduce the previous test errors in garbage collection.  Please merge.

glyph committed:
Merge branch 'trunk' into port-pb-7598

glyph commented:
I'm a _little_ unhappy with my own test coverage here, but I think that will actually be easier to address after landing this, so I'm going to go with @rodrigc's code review :-).  (Presuming required statuses pass...)

glyph commented:
stopping the OS X travis build because https://www.traviscistatus.com/incidents/4mvp857qx8bw

glyph commented:
Augh, cancelling the build, even though it's in `allow_failures`, causes the build to fail.  Everything else has passed though so I'm going to push the big _red_ button.

Ugh.

rodrigc commented:
@glyph thanks for porting PB to Py3.  That has opened the door to me submitting a lot of Py3 fixes upstream to buildbot: https://lists.buildbot.net/pipermail/devel/2017-January/012304.html

glyph commented:
@rodrigc -

> thanks for porting buildbot to Py3

I think you mean "thanks for porting _PB_ to Py3"? :)  If so - you're welcome, glad to help!

rodrigc commented:
Yeah I meant PB!
